client/matchmaking/admin-matchmaking-times.tsx

```diff
@@ -0,0 +1,690 @@
+  justify-content: space-between;
+  align-items: center;
+  padding: 16px;
+  background-color: rgba(255, 255, 255, 0.05);
+  border-radius: 8px;
+  transition: background-color 0.2s ease;
+
+  &:hover {
+    background-color: rgba(255, 255, 255, 0.08);
+  }
+`
+
+const CurrentMatchmakingCard = styled(HistoryCard)`
+  background-color: rgba(255, 193, 7, 0.1);
+  border: 1px solid rgba(255, 193, 7, 0.2);
+  box-shadow: 0 0 10px rgba(255, 193, 7, 0.1);
+
+  &:hover {
+    background-color: rgba(255, 193, 7, 0.15);
+  }
+`
+
+const HistoryInfo = styled.div`
+  display: flex;
+  flex-direction: column;
+  gap: 4px;
+`
+
+const HistoryDate = styled.div`
+  ${bodyMedium};
+  color: var(--theme-on-surface-variant);
+`
+
+const HistoryStatus = styled.div`
+  display: flex;
+  align-items: center;
+  gap: 8px;
+`
+
+const HistoryActions = styled.div`
+  display: flex;
+  gap: 8px;
+`
+
+const LoadMoreButton = styled(TextButton)`
+  width: 100%;
+  margin: 8px 0;
+`
+
+const EnabledText = styled.span`
+  color: var(--theme-success);
+`
+
+const DisabledText = styled.span`
+  color: var(--theme-error);
+`
+
+const CurrentText = styled.span`
+  color: var(--theme-amber);
+`
+
+const FinishedText = styled.span`
+  color: var(--theme-on-surface-variant);
+`
+
+const EmptyStateContainer = styled.div`
+  display: flex;
+  flex-direction: column;
+  align-items: center;
+  justify-content: center;
+  padding: 32px;
+  background-color: rgba(255, 255, 255, 0.03);
+  border-radius: 8px;
+  margin: 16px 0;
+  text-align: center;
+`
+
+const EmptyStateIcon = styled(MaterialIcon)`
+  font-size: 48px;
+  color: var(--theme-on-surface-variant);
+  margin-bottom: 24px;
+  opacity: 0.7;
+`
+
+const EmptyStateText = styled.p`
+  ${bodyMedium};
+  color: var(--theme-on-surface-variant);
+  max-width: 400px;
+  margin: 0;
+`
+
+export function AdminMatchmakingTimes() {
+  const { t } = useTranslation()
+
+  const [activeTab, setActiveTab] = useState<MatchmakingType>(MatchmakingType.Match1v1)
+
+  return (
+    <Container>
+      <Tabs activeTab={activeTab} onChange={setActiveTab}>
+        {ALL_MATCHMAKING_TYPES.map(type => (
+          <TabItem key={type} text={matchmakingTypeToLabel(type, t)} value={type} />
+        ))}
+      </Tabs>
+
+      <PageHeadline>Matchmaking times</PageHeadline>
+
+      {/** Recreate this component when the active tab changes so it clears its local state. */}
+      <AddAndDisplayMatchmakingTimes key={activeTab} activeTab={activeTab} />
+    </Container>
+  )
+}
+
+function AddAndDisplayMatchmakingTimes({ activeTab }: { activeTab: MatchmakingType }) {
+  const dispatch = useAppDispatch()
+  const snackbarController = useSnackbarController()
+
+  const [startDate, setStartDate] = useState('')
+  const [invalidDate, setInvalidDate] = useState(false)
+  const [enabled, setEnabled] = useState(false)
+
+  const [addError, setAddError] = useState<Error>()
+
+  const [futureTimesIds, setFutureTimesIds] = useImmerState<Set<number>>(new Set())
```

Comment:
I think this is overcomplicating things, just keep an array of the objects in a `useState`? What does a Set/Map provide us here? We're just iterating them to render anyway. Iterating 10s of items to find the right one to delete is straightforward/fast

Reply:
Actually I didn't do this primarily for the performance. But rather, to get around the annoying edge case of sometimes possibly having duplicates in the future times (e.g. we insert a new time into the array when adding it, and then also insert it when loading a new page if the added time is in that page). We were handling this manually previously (also by utilizing Sets), but thought separating the states like this makes it a bit "cleaner".
https://github.com/ShieldBattery/ShieldBattery/blob/2e120c81127c0d2b171ccc3ffad7d961aadad936/client/admin/matchmaking-times-reducer.js#L89-L90


I guess I could do the same thing again and just use a Set of objects? Or use an array and just do `Array.from(new Set(...))` to deduplicate?

Reply:
Probably just:
```ts
const seen = new Set()
return existingTimes.concat(futureTimes).filter(t => {
  const result = seen.has(t.id)
  seen.add(t.id)
  return result
})
```

but this isn't really correct either (and your old code also wasn't). If you retrieve an ordering from the server then that ordering should be assumed to be the correct one, not whatever you had locally. You can of course just sort by date or whatever in t his case, but yeah.